storage/src/nodestore.rs

```diff
@@ -587,56 +588,46 @@ pub trait HashedNodeReader: TrieReader {
     }
 }
 
-impl<R: RootReader> RootReader for Arc<R> {
-    fn root_node(&self) -> Option<Arc<Node>> {
-        R::root_node(self)
-    }
-}
-
-impl<T> TrieReader for Arc<T> where T: TrieReader {}
-
-impl<T> NodeReader for Arc<T>
+impl<T> HashedNodeReader for T
 where
-    T: NodeReader,
+    T: Deref,
+    T::Target: HashedNodeReader,
 {
-    fn read_node(&self, addr: LinearAddress) -> Result<Arc<Node>, Error> {
-        T::read_node(self, addr)
-    }
-}
-
-impl<H: HashedNodeReader> HashedNodeReader for Arc<H> {
     fn root_address_and_hash(&self) -> Result<Option<(LinearAddress, TrieHash)>, Error> {
-        H::root_address_and_hash(self)
-    }
-    fn root_hash(&self) -> Result<Option<TrieHash>, Error> {
-        H::root_hash(self)
+        self.deref().root_address_and_hash()
     }
 }
 
 /// Reads nodes and the root address from a merkle trie.
 pub trait TrieReader: NodeReader + RootReader {}
+impl<T> TrieReader for T where T: NodeReader + RootReader {}
+
+/// Reads nodes from a merkle trie.
+pub trait NodeReader {
+    /// Returns the node at `addr`.
+    fn read_node(&self, addr: LinearAddress) -> Result<Arc<Node>, Error>;
+}
 
-impl TrieReader for &NodeStore<Committed, FileBacked> {}
-impl NodeReader for &NodeStore<Committed, FileBacked> {
+impl<T> NodeReader for T
+where
+    T: Deref,
+    T::Target: NodeReader,
+{
     fn read_node(&self, addr: LinearAddress) -> Result<Arc<Node>, Error> {
-        self.read_node_from_disk(addr)
+        self.deref().read_node(addr)
     }
 }
 
-impl RootReader for &NodeStore<Committed, FileBacked> {
+impl<T> RootReader for T
+where
+    T: Deref,
+    T::Target: RootReader,
+{
     fn root_node(&self) -> Option<Arc<Node>> {
-        self.header
-            .root_address
-            .map(|addr| self.read_node_from_disk(addr).unwrap())
```

Comment:
This might be wrong. If that's the case, you can just close this PR. If it's wrong though, that means you've found a bit of a hack for specialization which is *a little* (not a lot) problematic for maintainability IMO. Usually specialization should come in the form of newtype wrappers, where you would end up with something like this:
```rust
node_store.into_specialized().root_node()
```
Where the specific implementations for that one function would differ. Then again, this might be okay? I would have to spend more time with it to figure that out.

Reply:
Discussed in chat. It's fine, although we should consider removing the whole NodeReader trait since reading nodes is always done the same way now.